aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -66,236 +57,344 @@ impl AggregatorID {
-                // change depending on how big the `value` is. Suppose
-                // aggregator has +X and want to do -Y. Then, there are two
-                // cases:
-                //     1. X >= Y: then the result is +(X-Y)
-                //     2. X  < Y: then the result is -(Y-X)
-                if self.value >= value {
-                    self.value = subtraction(self.value, value)?;
+            AggregatorState::Delta {
+                speculative_start_value,
+                delta,
+                history,
+            } => {
+                let cur_value = expect_ok(math.unsigned_add_delta(
+                    speculative_start_value.get_any_value()?,
+                    delta,
+                ))?;
+
+                if cur_value < input {
+                    let underflow_delta = expect_ok(ok_overflow(math.unsigned_add_delta(
+                        input,
+                        &delta.minus(),
+                    )))?;
+                    // If value overflowed (delta was smaller than -max_value), we don't need to record it.
+                    if let Some(underflow_delta) = underflow_delta {
+                        history.record_underflow(underflow_delta);
+                    }
+                    Ok(false)
                 } else {
-                    // Check that we can subtract in general: we don't want to
-                    // allow -10000 when limit is 10.
-                    // TODO: maybe `subtraction` should also know about the limit?
-                    subtraction(self.limit, value)?;
-
-                    self.value = subtraction(value, self.value)?;
-                    self.state = AggregatorState::NegativeDelta;
+                    let new_delta =
+                        expect_ok(math.signed_add(delta, &SignedU128::Negative(input)))?;
+                    *delta = new_delta;
+                    history.record_success(new_delta);
+                    Ok(true)
                 }
             },
-            AggregatorState::NegativeDelta => {
-                // Since we operate on unsigned integers, we have to add
-                // when subtracting from negative delta. Note that if limit
-                // is some X, then we cannot subtract more than X, and so
-                // we should return an error there.
-                self.value = addition(self.value, value, self.limit)?;
-            },
         }
+    }
 
-        // Record side-effects of addition in history.
-        self.record();
+    /// Implements logic for doing a "cheap read" of an aggregator.
+    /// Reads the last committed value of the aggregator that's known at the
+    /// time of the call, and as such, can be computed efficiently (i.e. no
+    /// need to consider any speculative state, deltas, etc)
+    /// This method has a sideffect, of updating `speculative_start_value` with
+    /// `LastCommittedValue` variant.
+    /// `get_any_value()` is guaranteed to succeed after this call.
```

Comment:
Should we add the invariant also, i.e. aggregator can't be a non-trivial delta state (pre-condition to initialize Delta state basically).